ibis/expr/operations/relations.py

```diff
@@ -96,6 +96,12 @@ def blocks(self):
         return True
 
 
+@public
+class TableLiteral(TableNode, sch.HasSchema):
```

Comment:
How is table literal different from a literal with datatype `dt.Array(dt.Struct(...))`?

Reply:
In principle nothing, but in practice there are issues dealing with nullable, the APIs are different, and many backends supports `VALUES` and do not support arrays or structs.

Reply:
Semantically there's also the issue that a relation and an array of structs are distinct entities.

Reply:
I'd rather like to see a bigger picture here providing a pluggable `ibis.table()` API, something like:

```python
@public
class TableNode(Node):
    def sort_by(self, sort_exprs):
        return Selection(self, [], sort_keys=sort_exprs)

    @property
    @abstractmethod
    def schema(self) -> sch.Schema:
        """Return a schema."""

    def to_expr(self):
        import ibis.expr.types as ir

        return ir.Table(self)


@public
class PhysicalTable(TableNode, Named):
    pass


@public
class UnboundTable(PhysicalTable):
    schema = rlz.instance_of(sch.Schema)
    name = rlz.optional(rlz.instance_of(str), default=genname)


@public
class DatabaseTable(PhysicalTable):
    name = rlz.instance_of(str)
    schema = rlz.instance_of(sch.Schema)
    source = rlz.client

    def change_name(self, new_name):
        return type(self)(new_name, self.args[1], self.source)


class InMemoryTable(PhysicalTable):
    pass


# implemented in the pandas backend
class PandasTable(InMemoryTable):
    data = rlz.hashable_wrapper_of(pd.DataFrame)

    @property
    def name(self):
        return self.data.name


# supported by duckdb, clickhouse, datafusion, acero
class S3Dataset(PhysicalTable):
    name = rlz.instance_of(str)
    pattern = rlz.instance_of(str)
    format = rlz.isin({"csv", "parquet", "json"})  # or whatever
```

Then a backend could register its own datatype as in `InMemoryTable`

```python
@ibis.table.register(pd.DataFrame):
def table_from_pandas(df, schema):
    # oversimplified impl for demonstration purposes
    return PandasTable(df, schema=schema)
```

Reply:
Another relevant operation is FileTable https://github.com/ibis-project/ibis/pull/4293

Reply:
I don't think `FileTable` is related to in memory tables. They are independent types of table scans.

Reply:
I mean they are related on the abstraction level, so if we are about to support in memory tables then we should consider how can we add support for additional table types.